README.md
# gosh-bls-lib

This library is developed to handle BLS signature for GOSH blockchain. It is responsible for generation BLS keys and provides basic signing/verification functionality. Also it provides aggregation functionality for public keys and signatures. 
Basic BLS functionality is provided by [blst](https://github.com/supranational/blst) library. _gosh-bls-lib_ exploits it and adds additional infrastracture and extra data parsing.

Raw BLS signature is a byte array of length 96 (or 48) bytes. Each BLS signature is produced initially by some node with unique index _node_index_. Also there is some fixed number of nodes for current validation session. Node concatenates BLS signature with _node_index_ and _total_number_of_nodes_. So it looks as follows.

<p align="center">
<img src="../master/bls_diag.jpg" width="600">
</p>

We need to add information about node indexes because in the end we (masterchain validator) should be able to compose appropriate aggregated BLS public key for verification of obtained (from workchain) BLS signature
  
So this wrapped BLS signature is broadcasted and aggregated with other BLS signatures for the same block candidate hash.
  
We need extra field for number of occurrences because we gonna use decentralized algorithm for BLS signatures broadcast and aggregation. So we may get the following situation. 
  
<p align="center">
<img src="../master/decentral.jpg" width="600">
</p>
  
When we will calculate aggregated public key to verify the final signature we should take into account how many times each node took part. So the final aggregated signature will look like this.
  
<p align="center">
<img src="../master/bls_diag_4.jpg" width="800">
</p>

When masterchain validator gets such aggregated BLS signature in broadcast protection message, he will parse it. He will take respective BLS public keys based on indexes in wrapped BLS signature. It will take into account the number of repetitions for each public key and compute the appropriate aggregation of public keys. Then it will verify the signature.

## API description

Below there are constants and functions that are provided by _gosh-bls-lib_.

```rust
pub const BLS_SECRET_KEY_LEN: usize = 32;
pub const BLS_PUBLIC_KEY_LEN_FOR_MIN_PK_MODE: usize = 48;
pub const BLS_PUBLIC_KEY_LEN_FOR_MIN_SIG_MODE: usize = 96;
pub const BLS_PUBLIC_KEY_LEN: usize = BLS_PUBLIC_KEY_LEN_FOR_MIN_PK_MODE;
pub const BLS_KEY_MATERIAL_LEN: usize = 32;
pub const BLS_SIG_LEN_FOR_MIN_PK_MODE: usize = 96;
pub const BLS_SIG_LEN_FOR_MIN_SIG_MODE: usize = 48;
pub const BLS_SIG_LEN: usize = BLS_SIG_LEN_FOR_MIN_PK_MODE;
pub const BLS_SEED_LEN: usize = 32;
```
- **gen_bls_key_pair_based_on_key_material**  
  
  ```rust 
  pub fn gen_bls_key_pair_based_on_key_material(ikm: &[u8; BLS_KEY_MATERIAL_LEN]) -> Result<([u8; BLS_PUBLIC_KEY_LEN], [u8; BLS_SECRET_KEY_LEN])> 
  ```
  Generate random BLS key pair based on key material array. Key material is not equal to future secret key in the end.

- **gen_bls_key_pair**  

  ```rust 
  pub fn gen_bls_key_pair() -> Result<([u8; BLS_PUBLIC_KEY_LEN], [u8; BLS_SECRET_KEY_LEN])>
  ```
  Generate random BLS key pair.
    
- **gen_public_key_based_on_secret_key**
  
   ```rust 
  pub fn gen_public_key_based_on_secret_key(sk: &[u8; BLS_SECRET_KEY_LEN]) -> Result<([u8; BLS_PUBLIC_KEY_LEN])>)>
  ```
  
  Generate public key using bytes of secret key.
  
- **sign** 

  ```rust 
  pub fn sign(sk_bytes: &[u8; BLS_SECRET_KEY_LEN], msg: &Vec<u8>) -> Result<[u8; BLS_SIG_LEN]>
  ```
  
  Compute raw 96/48-bytes BLS signature.

- **verify**

  ```rust 
  pub fn verify(sig_bytes: &[u8; BLS_SIG_LEN], msg: &Vec<u8>, pk_bytes: &[u8; BLS_PUBLIC_KEY_LEN]) -> Result<bool>
  ```
  
  Verify raw BLS signature.
  
- **add_node_info_to_sig**
  
  ```rust 
  pub fn add_node_info_to_sig(sig_bytes: [u8; BLS_SIG_LEN], node_index: u16, total_num_of_nodes: u16) -> Result<Vec<u8>>
  ```
  
  Concatenate raw BLS signature bytes with node index and total number of nodes.
  
- **sign_and_add_node_info**
  
  ```rust 
    pub fn sign_and_add_node_info(sk_bytes: &[u8; BLS_SECRET_KEY_LEN], msg: &Vec<u8>, node_index: u16, total_num_of_nodes: u16) -> Result<Vec<u8>>
  ```
  
  Create raw BLS signature and concatenate it with node index and total number of nodes.
  
- **truncate_nodes_info_from_sig**

  ```rust 
    pub fn truncate_nodes_info_from_sig(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<[u8; BLS_SIG_LEN]>
  ```
  
  Truncare raw BLS signature bytes from wrapped signature.
  
- **get_nodes_info_from_sig**
  
  ```rust 
    pub fn get_nodes_info_from_sig(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<Vec<u8>>
  ```
  Truncate info about nodes (indexes, number of occurrences, total number of nodes) from wrapped BLS signature.
  
- **truncate_nodes_info_and_verify**

  ```rust 
    pub fn truncate_nodes_info_and_verify(sig_bytes_with_nodes_info: &Vec<u8>, pk_bytes: &[u8; BLS_PUBLIC_KEY_LEN], msg: &Vec<u8>) ->Result<bool>
  ```

  Truncate info about nodes (indexes, number of occurrences, total number of nodes) from wrapped BLS signature and then verify raw signature.

- **aggregate_public_keys**

  ```rust 
    pub fn aggregate_public_keys(bls_pks_bytes: &Vec<&[u8; BLS_PUBLIC_KEY_LEN]>) -> Result<[u8; BLS_PUBLIC_KEY_LEN]>
  ```
  
  Aggregate all public keys taken from vector.

- **aggregate_public_keys_based_on_nodes_info**

  ```rust 
    pub fn aggregate_public_keys_based_on_nodes_info(bls_pks_bytes: &Vec<&[u8; BLS_PUBLIC_KEY_LEN]>, nodes_info_bytes: &Vec<u8>) -> Result<[u8; BLS_PUBLIC_KEY_LEN]> 
  ```
  
  Aggregate public keys based on nodes_info map containing indexes of nodes. Here node_info_bytes has the same structure as described before: first 2 bytes contains total number of nodes and then array of pairs [index (2 bytes), number of occurrences (2 bytes)].

- **aggregate_two_bls_signatures**

  ```rust 
    pub fn aggregate_two_bls_signatures(bls_sig_1_bytes: &Vec<u8>, bls_sig_2_bytes: &Vec<u8>) -> Result<Vec<u8>> 
  ```
  
  Aggregate two BLS signatures and merge their node info.
  
- **aggregate_bls_signatures**

  ```rust 
    pub fn aggregate_bls_signatures(bls_sigs_bytes: &Vec<&Vec<u8>>) -> Result<Vec<u8>> 
  ```
  
  Aggregate multiple BLS signatures and merge their node info.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Version: 0.1.24

### New
 - Update Cargo.toml





src/bls/sig.rs
use super::converters::*;
use super::key_gen::*;
use super::nodes_info::*;
use super::random_helper::*;
use std::convert::TryInto;
use blst::*;

use blst::min_pk::*;
//use blst::min_sig::*;

use ton_types::{fail, Result};

use crate::bls::{add_node_info_to_sig, BLS_PUBLIC_KEY_LEN, BLS_SIG_LEN};
use crate::bls::BLS_SECRET_KEY_LEN;
use crate::bls::BLS_KEY_MATERIAL_LEN;

pub const DST: [u8; 43] = *b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_";

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct BlsSignature {
    pub sig_bytes: [u8; BLS_SIG_LEN],
    pub nodes_info: NodesInfo,
}

impl BlsSignature {
    pub fn serialize(&self) -> Vec<u8> {
        let mut vec = Vec::new();
        vec.extend_from_slice(&self.sig_bytes);
        let nodes_info_bytes = &self.nodes_info.serialize();
        vec.extend_from_slice(&nodes_info_bytes);
        vec
    }

    pub fn deserialize(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<Self> {
        if sig_bytes_with_nodes_info.len() < BLS_SIG_LEN + 6 {
            fail!("Length of sig_bytes_with_nodes_info is too short!")
        }
        let mut sig_bytes: [u8; BLS_SIG_LEN] = [0; BLS_SIG_LEN];
        sig_bytes.copy_from_slice(&sig_bytes_with_nodes_info[0..BLS_SIG_LEN]);
        let len = sig_bytes_with_nodes_info.len() - BLS_SIG_LEN;
        let mut nodes_info_data = vec![0; len];
        nodes_info_data.copy_from_slice(&sig_bytes_with_nodes_info[BLS_SIG_LEN..]);
        let nodes_info = NodesInfo::deserialize(&nodes_info_data)?;
        Ok(Self{sig_bytes, nodes_info})
    }

    pub fn simple_sign(sk_bytes: &[u8; BLS_SECRET_KEY_LEN], msg: &Vec<u8>) -> Result<[u8; BLS_SIG_LEN]> {
        if msg.len() == 0 {
            fail!("Msg to sign can not be empty!")
        }
        let sk = convert_secret_key_bytes_to_secret_key(sk_bytes)?;
        let sig = sk.sign(msg, &DST, &[]);
        Ok(sig.to_bytes())
    }

    pub fn simple_verify(sig_bytes: &[u8; BLS_SIG_LEN], msg: &Vec<u8>, pk_bytes: &[u8; BLS_PUBLIC_KEY_LEN]) -> Result<bool> {
        if msg.len() == 0 {
            fail!("Msg to sign can not be empty!")
        }
        let sig = convert_signature_bytes_to_signature(sig_bytes)?;
        let pk = convert_public_key_bytes_to_public_key(pk_bytes)?;
        let res = sig.verify(true, msg, &DST, &[], &pk, true);
        Ok(res == BLST_ERROR::BLST_SUCCESS)
    }

    pub fn add_node_info_to_sig(sig_bytes: [u8; BLS_SIG_LEN], node_index: u16, total_num_of_nodes: u16) -> Result<Vec<u8>> {
        if total_num_of_nodes == 0 {
            fail!("Total number of nodes can not be zero!");
        }
        if node_index >= total_num_of_nodes {
            fail!("Index of node can not be greater than total number of nodes!");
        }
        let nodes_info = NodesInfo::create_node_info(total_num_of_nodes, node_index)?;
        let sig = Self {
            sig_bytes,
            nodes_info,
        };
        let sig_bytes = BlsSignature::serialize(&sig);
        Ok(sig_bytes)
    }

    pub fn sign(
        sk_bytes: &[u8; BLS_SECRET_KEY_LEN],
        msg: &Vec<u8>,
        node_index: u16,
        total_num_of_nodes: u16,
    ) -> Result<Vec<u8>> {
        let sig = BlsSignature::simple_sign(sk_bytes, msg)?;
        add_node_info_to_sig(sig, node_index, total_num_of_nodes)
    }

    pub fn get_nodes_info_from_sig(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<Vec<u8>> {
        let bls_sig = BlsSignature::deserialize(sig_bytes_with_nodes_info)?;
        Ok(bls_sig.nodes_info.serialize())
    }

    pub fn truncate_nodes_info_from_sig(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<[u8; BLS_SIG_LEN]> {
        let bls_sig = BlsSignature::deserialize(sig_bytes_with_nodes_info)?;
        Ok(bls_sig.sig_bytes)
    }

    pub fn verify(sig_bytes_with_nodes_info: &Vec<u8>, pk_bytes: &[u8; BLS_PUBLIC_KEY_LEN], msg: &Vec<u8>) -> Result<bool> {
        let sig_bytes = BlsSignature::truncate_nodes_info_from_sig(sig_bytes_with_nodes_info)?;
        let res = BlsSignature::simple_verify(&sig_bytes, msg, pk_bytes)?;
        Ok(res)
    }

    pub fn print_signature_bytes(sig_bytes: &[u8]) {
        if sig_bytes.len() != BLS_SIG_LEN {
            panic!("Incorrect length of signature byte array!")
        }
        println!("--------------------------------------------------");
        println!("BLS Signature bytes:");
        println!("--------------------------------------------------");
        println!("{:?}", sig_bytes);
        println!("--------------------------------------------------");
    }

    pub fn print_bls_signature(bls_sig_bytes: &Vec<u8>) {
        let bls_sig = BlsSignature::deserialize(bls_sig_bytes).unwrap();
        bls_sig.print();
    }

    pub fn print(&self) {
        println!("--------------------------------------------------");
        println!("Aggregated BLS signature:");
        println!("--------------------------------------------------");
        println!("Signature bytes:");
        println!("{:?}", &self.sig_bytes);
        &self.nodes_info.print();
        println!("--------------------------------------------------");
    }
}




src/bls/key_gen.rs
use super::converters::*;
use blst::*;
use blst::min_pk::*;
//use blst::min_sig::*;
use rand::Rng;
use rand::{RngCore};

use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

use std::collections::HashSet;

use ton_types::{fail, Result};

use crate::bls::BLS_PUBLIC_KEY_LEN;
use crate::bls::BLS_SECRET_KEY_LEN;
use crate::bls::BLS_KEY_MATERIAL_LEN;

pub struct KeyPair {
    pub sk: SecretKey,
    pub pk: PublicKey,
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct BlsKeyPair {
    pub pk_bytes: [u8; BLS_PUBLIC_KEY_LEN],
    pub sk_bytes: [u8; BLS_SECRET_KEY_LEN]
}

impl BlsKeyPair {
    pub fn print_bls_public_key(bls_pk_bytes: &[u8]) {
        if bls_pk_bytes.len() != BLS_PUBLIC_KEY_LEN{
            panic!("Incorrect length of secret key byte array!")
        }
        println!("--------------------------------------------------");
        println!("Aggregated BLS public key");
        println!("--------------------------------------------------");
        println!("Public key bytes:");
        println!("{:?}", bls_pk_bytes);
        println!("--------------------------------------------------");
    }

    pub fn print(&self) {
        println!("--------------------------------------------------");
        println!("BLS key pair:");
        println!("--------------------------------------------------");
        println!("Secret key bytes:");
        println!("{:?}", &self.sk_bytes);
        println!("Secret key len: {}", &self.sk_bytes.len());
        println!("Public key bytes:");
        println!("{:?}", &self.pk_bytes);
        println!("Public key len: {}", &self.pk_bytes.len());
        println!("--------------------------------------------------");
    }


    pub fn serialize(&self) -> ([u8; BLS_PUBLIC_KEY_LEN], [u8; BLS_SECRET_KEY_LEN])  {
        (self.pk_bytes, self.sk_bytes)
    }

    pub fn deserialize(key_pair_data: &([u8; BLS_PUBLIC_KEY_LEN], [u8; BLS_SECRET_KEY_LEN])) -> Result<Self> {
        let sk = convert_secret_key_bytes_to_secret_key(&key_pair_data.1)?;
        let pk = sk.sk_to_pk();
        if key_pair_data.0 != pk.to_bytes() {
            fail!("Public key does not correspond to secret key!")
        }
        Ok(Self {
            pk_bytes: key_pair_data.0,
            sk_bytes: key_pair_data.1
        })
    }

    pub fn deserialize_based_on_secret_key(sk_bytes: &[u8; BLS_SECRET_KEY_LEN]) -> Result<Self> {
        let sk = convert_secret_key_bytes_to_secret_key(sk_bytes)?;
        let pk = sk.sk_to_pk();
        Ok(Self {
            pk_bytes: pk.to_bytes(),
            sk_bytes: sk.to_bytes()
        })
    }

    pub fn gen_bls_key_pair_based_on_key_material(ikm: &[u8; BLS_KEY_MATERIAL_LEN]) -> Result<Self> {
        let key_pair = BlsKeyPair::gen_key_pair_based_on_key_material(&ikm)?;
        Ok(BlsKeyPair::convert_key_pair_to_bls_key_pair(key_pair))
    }

    pub fn gen_bls_key_pair() -> Result<Self> {
        let key_pair = BlsKeyPair::gen_key_pair()?;
        Ok(BlsKeyPair::convert_key_pair_to_bls_key_pair(key_pair))
    }

    pub fn gen_key_pair() -> Result<KeyPair> {
        let mut ikm = [0u8; BLS_KEY_MATERIAL_LEN];
        rand::thread_rng().fill_bytes(&mut ikm);
        BlsKeyPair::gen_key_pair_based_on_key_material(&ikm)
    }

    pub fn gen_key_pair_based_on_key_material(ikm: &[u8; BLS_KEY_MATERIAL_LEN]) -> Result<KeyPair> {
        if ikm.len() != BLS_KEY_MATERIAL_LEN {
            fail!("Incorrect length of key material byte array!")
        }
        if let Ok(sk) = SecretKey::key_gen(ikm, &[]) {
            let pk = sk.sk_to_pk();
            Ok(KeyPair { sk: sk, pk: pk })
        } else {
            fail!("Failed while generate key")
        }
    }

    fn convert_key_pair_to_bls_key_pair(key_pair: KeyPair) -> Self {
        return BlsKeyPair {
            sk_bytes: key_pair.sk.to_bytes(),
            pk_bytes: key_pair.pk.to_bytes(),
        };
    }
}















src/bls.rs

mod key_gen;
mod sig;
mod nodes_info;
mod converters;
mod aggregate;
mod random_helper;

pub use self::key_gen::*;
pub use self::sig::*;
pub use self::aggregate::*;
pub use self::nodes_info::*;
pub use self::random_helper::*;
use blst::*;

use blst::min_pk::*;
//use blst::min_sig::*;

use rand::Rng;
use rand::RngCore;

use ton_types::{fail, Result};
use std::time::{Instant, Duration};
use std::convert::TryInto;
use crate::bls::random_helper::{generate_random_msg, generate_random_msg_of_fixed_len};

pub const BLS_SECRET_KEY_LEN: usize = 32;
pub const BLS_PUBLIC_KEY_LEN_FOR_MIN_PK_MODE: usize = 48;
pub const BLS_PUBLIC_KEY_LEN_FOR_MIN_SIG_MODE: usize = 96;
pub const BLS_PUBLIC_KEY_LEN: usize = BLS_PUBLIC_KEY_LEN_FOR_MIN_PK_MODE;
pub const BLS_KEY_MATERIAL_LEN: usize = 32;
pub const BLS_SIG_LEN_FOR_MIN_PK_MODE: usize = 96;
pub const BLS_SIG_LEN_FOR_MIN_SIG_MODE: usize = 48;
pub const BLS_SIG_LEN: usize = BLS_SIG_LEN_FOR_MIN_PK_MODE;
pub const BLS_SEED_LEN: usize = 32;


pub fn gen_bls_key_pair_based_on_key_material(ikm: &[u8; BLS_KEY_MATERIAL_LEN]) -> Result<([u8; BLS_PUBLIC_KEY_LEN], [u8; BLS_SECRET_KEY_LEN])> {
    let key_pair = BlsKeyPair::gen_bls_key_pair_based_on_key_material(ikm)?;
    Ok(key_pair.serialize())
}

pub fn gen_bls_key_pair() -> Result<([u8; BLS_PUBLIC_KEY_LEN], [u8; BLS_SECRET_KEY_LEN])> {
    let key_pair = BlsKeyPair::gen_bls_key_pair()?;
    Ok(key_pair.serialize())
}

pub fn gen_public_key_based_on_secret_key(sk: &[u8; BLS_SECRET_KEY_LEN]) -> Result<([u8; BLS_PUBLIC_KEY_LEN])> {
    let pk = BlsKeyPair::deserialize_based_on_secret_key(sk)?;
    Ok(pk.pk_bytes)
}

pub fn sign(sk_bytes: &[u8; BLS_SECRET_KEY_LEN], msg: &Vec<u8>) -> Result<[u8; BLS_SIG_LEN]>  {
    BlsSignature::simple_sign(sk_bytes, msg)
}

pub fn verify(sig_bytes: &[u8; BLS_SIG_LEN], msg: &Vec<u8>, pk_bytes: &[u8; BLS_PUBLIC_KEY_LEN]) -> Result<bool> {
    BlsSignature::simple_verify(sig_bytes, msg, pk_bytes)
}

pub fn add_node_info_to_sig(sig_bytes: [u8; BLS_SIG_LEN], node_index: u16, total_num_of_nodes: u16) -> Result<Vec<u8>> {
    BlsSignature::add_node_info_to_sig(sig_bytes, node_index, total_num_of_nodes)
}

pub fn sign_and_add_node_info(
    sk_bytes: &[u8; BLS_SECRET_KEY_LEN],
    msg: &Vec<u8>,
    node_index: u16,
    total_num_of_nodes: u16,
) -> Result<Vec<u8>> {
    BlsSignature::sign(sk_bytes, msg, node_index, total_num_of_nodes)
}

pub fn truncate_nodes_info_from_sig(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<[u8; BLS_SIG_LEN]> {
    BlsSignature::truncate_nodes_info_from_sig(sig_bytes_with_nodes_info)
}

pub fn get_nodes_info_from_sig(sig_bytes_with_nodes_info: &Vec<u8>) -> Result<Vec<u8>> {
    BlsSignature::get_nodes_info_from_sig(sig_bytes_with_nodes_info)
}

pub fn truncate_nodes_info_and_verify(sig_bytes_with_nodes_info: &Vec<u8>, pk_bytes: &[u8; BLS_PUBLIC_KEY_LEN], msg: &Vec<u8>) -> Result<bool> {
    BlsSignature::verify(sig_bytes_with_nodes_info, pk_bytes, msg)
}

pub fn aggregate_bls_signatures(bls_sigs_bytes: &Vec<&Vec<u8>>) -> Result<Vec<u8>> {
    aggregate::aggregate_bls_signatures(bls_sigs_bytes)
}

pub fn aggregate_two_bls_signatures(bls_sig_1_bytes: &Vec<u8>, bls_sig_2_bytes: &Vec<u8>) -> Result<Vec<u8>> {
    aggregate::aggregate_two_bls_signatures(bls_sig_1_bytes, bls_sig_2_bytes)
}

pub fn aggregate_public_keys_based_on_nodes_info(bls_pks_bytes: &Vec<&[u8; BLS_PUBLIC_KEY_LEN]>, nodes_info_bytes: &Vec<u8>) -> Result<[u8; BLS_PUBLIC_KEY_LEN]> {
    aggregate::aggregate_public_keys_based_on_nodes_info(bls_pks_bytes, nodes_info_bytes)
}

pub fn aggregate_public_keys(bls_pks_bytes: &Vec<&[u8; BLS_PUBLIC_KEY_LEN]>) -> Result<[u8; BLS_PUBLIC_KEY_LEN]> {
    aggregate::aggregate_public_keys(bls_pks_bytes)
}

pub fn print_bls_public_key(bls_pk_bytes: &[u8]) {
    BlsKeyPair::print_bls_public_key(bls_pk_bytes)
}

pub fn print_signature_bytes(sig_bytes: &[u8]) {
    BlsSignature::print_signature_bytes(sig_bytes)
}

pub fn print_bls_signature(bls_sig_bytes: &Vec<u8>) {
    BlsSignature::print_bls_signature(bls_sig_bytes)
}


#[test]
fn test_gen_bls_key_pair() {
    for i in 0..100 {
        let now = Instant::now();
        let key_pair = gen_bls_key_pair().unwrap();
        let duration = now.elapsed();
      //  println!("Public key : {:?}", key_pair.0);
       // println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by gen_bls_key_pair is: {:?}", duration);
    }
}

#[test]
fn test_gen_bls_key_pair_based_on_key_material() {
    let mut ikm = [0u8; BLS_KEY_MATERIAL_LEN];
    for i in 0..100 {
        let now = Instant::now();
        let key_pair = gen_bls_key_pair_based_on_key_material(&ikm).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by gen_bls_key_pair_based_on_key_material is: {:?}", duration);
    }
}

#[test]
fn test_gen_public_key_based_on_secret_key() {
    for i in 0..100 {
        let key_pair = gen_bls_key_pair().unwrap();
        let now = Instant::now();
        let pk = gen_public_key_based_on_secret_key(&key_pair.1).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by gen_public_key_based_on_secret_key is: {:?}", duration);
        assert_eq!(pk, key_pair.0);
    }
}

#[test]
fn test_sign() {
    for i in 0..100 {
        let key_pair = gen_bls_key_pair().unwrap();
        let msg = generate_random_msg_of_fixed_len(10000000);
        let now = Instant::now();
        let sig = sign(&key_pair.1, &msg).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by sign is: {:?}", duration);
       // assert_eq!(pk, key_pair.0);
    }
}

#[test]
fn test_verify() {
    for i in 0..100 {
        let key_pair = gen_bls_key_pair().unwrap();
        let msg = generate_random_msg_of_fixed_len(1000);
        let sig = sign(&key_pair.1, &msg).unwrap();
        let now = Instant::now();
        let res = verify(&sig, &msg, &key_pair.0).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by verify is: {:?}", duration);
         assert_eq!(res, true);
    }
}

#[test]
fn test_add_node_info_to_sig() {
    let index = 100;
    let total_num_of_index = 10000;
    for i in 0..100 {
        let key_pair = gen_bls_key_pair().unwrap();
        let msg = generate_random_msg_of_fixed_len(500000);
        let sig = sign(&key_pair.1, &msg).unwrap();
        let now = Instant::now();
        let res = add_node_info_to_sig(sig, index, total_num_of_index).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by add_node_info_to_sig is: {:?}", duration);
    }
}

#[test]
fn test_sign_and_add_node_info() {
    let index = 100;
    let total_num_of_index = 1000;
    for i in 0..100 {
        let key_pair = gen_bls_key_pair().unwrap();
        let msg = generate_random_msg_of_fixed_len(10000000);
        let now = Instant::now();
        let res = sign_and_add_node_info(&key_pair.1, &msg, index, total_num_of_index).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by sign_and_add_node_info is: {:?}", duration);
    }
}

#[test]
fn test_aggregate_public_keys() {
    let number_of_keys = 10000;
    for i in 0..10 {
        let mut public_keys = Vec::new();
        for j in 0..number_of_keys {
            let key_pair = gen_bls_key_pair().unwrap();
            public_keys.push(key_pair.0);
        }
        let public_keys_refs: Vec<&[u8; BLS_PUBLIC_KEY_LEN]> = public_keys.iter().map(|pk| pk).collect();
        let now = Instant::now();
        let res = aggregate_public_keys(&public_keys_refs).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by aggregate_public_keys is: {:?}", duration);
    }
}

#[test]
fn test_aggregate_public_keys_based_on_nodes_info() {
    let total_num_of_nodes = 10000;
    for i in 0..10 {
        let indexes: Vec<u16> =  gen_signer_indexes(total_num_of_nodes, total_num_of_nodes * 2);
        let mut node_info_vec = Vec::new();
        for ind in &indexes {
            //println!("Node index = {}", ind);
            let nodes_info = NodesInfo::create_node_info(total_num_of_nodes, *ind).unwrap();
            node_info_vec.push(nodes_info)

        }
        let node_info_vec_refs: Vec<&NodesInfo> = node_info_vec.iter().map(|info| info).collect();
        let info = NodesInfo::merge_multiple(&node_info_vec_refs).unwrap();
        println!("Node info size = {}", info.map.len());
       // info.print();

        let mut public_keys = Vec::new();
        for j in 0..total_num_of_nodes {
            let key_pair = gen_bls_key_pair().unwrap();
            public_keys.push(key_pair.0);
        }
        let public_keys_refs: Vec<&[u8; BLS_PUBLIC_KEY_LEN]> = public_keys.iter().map(|pk| pk).collect();
        let now = Instant::now();
        let res = aggregate_public_keys_based_on_nodes_info(&public_keys_refs, &info.serialize()).unwrap();
        let duration = now.elapsed();

        println!("Time elapsed by aggregate_public_keys is: {:?}", duration);
    }
}

#[test]
fn test_aggregate_two_bls_signatures() {
    let number_of_keys = 100;
    for i in 0..10 {
        let key_pair_1 = gen_bls_key_pair().unwrap();
        let key_pair_2 = gen_bls_key_pair().unwrap();
        let msg = generate_random_msg();
        let ind_1 = gen_random_index(number_of_keys);
        let ind_2 = gen_random_index(number_of_keys);
        let sig_1 = sign_and_add_node_info(&key_pair_1.1, &msg, ind_1, number_of_keys).unwrap();
        let sig_2 = sign_and_add_node_info(&key_pair_2.1, &msg, ind_2, number_of_keys).unwrap();
        let now = Instant::now();
        let res = aggregate_two_bls_signatures(&sig_1, &sig_2).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by aggregate_two_bls_signatures is: {:?}", duration);
    }
}

#[test]
fn test_aggregate_two_bls_signatures_2() {
    let number_of_keys = 10000;
    for i in 0..10 {
        let key_pair_1 = gen_bls_key_pair().unwrap();
        let key_pair_2 = gen_bls_key_pair().unwrap();
        let msg = generate_random_msg();

        let sig_1 = sign(&key_pair_1.1, &msg).unwrap();
        let sig_2 = sign(&key_pair_2.1, &msg).unwrap();
        let info_1 = create_random_nodes_info(number_of_keys, number_of_keys * 2);
        let info_2 = create_random_nodes_info(number_of_keys, number_of_keys * 2);
        println!("info_1 size: {:?}", &info_1.map.len());

        let bls_sig_1 =  BlsSignature {
            sig_bytes: sig_1,
            nodes_info: info_1
        }.serialize();

        println!("info_2 size: {:?}", &info_2.map.len());

        let bls_sig_2 =  BlsSignature {
            sig_bytes: sig_2,
            nodes_info: info_2
        }.serialize();

        let now = Instant::now();
        let res = aggregate_two_bls_signatures(&bls_sig_1, &bls_sig_2).unwrap();
        let duration = now.elapsed();
        //  println!("Public key : {:?}", key_pair.0);
        //println!("Secret key : {:?}", key_pair.1);
        println!("Time elapsed by aggregate_two_bls_signatures is: {:?}", duration);
    }
}

#[test]
fn test_aggregate_bls_signatures() {
    let number_of_keys = 10000;
    let number_of_signatures = 50;
    for i in 0..10 {
        let mut sigs = Vec::new();
        let msg = generate_random_msg();
        for j in 0..number_of_signatures {
            let key_pair = gen_bls_key_pair().unwrap();
            let sig = sign(&key_pair.1, &msg).unwrap();
            let info = create_random_nodes_info(number_of_keys, number_of_keys * 2);
            println!("info size: {:?}", &info.map.len());
            let bls_sig =  BlsSignature {
                sig_bytes: sig,
                nodes_info: info
            }.serialize();
            sigs.push(bls_sig);
        }
        let sigs_refs: Vec<&Vec<u8>> = sigs.iter().map(|sig| sig).collect();

        let now = Instant::now();
        let res = aggregate_bls_signatures(&sigs_refs).unwrap();
        let duration = now.elapsed();

        println!("Time elapsed by aggregate_bls_signatures is: {:?}", duration);
    }
}
